File: workers/src/outcome.js
/**
 * Job outcomes.
 *
 * A worker handler does not touch the SDK job object directly. It returns one
 * of the outcomes below, which keeps the business logic testable on its own and
 * keeps the choice of job action in one place.
 *
 * The three outcomes map onto the three ways a job can end in Camunda 8:
 *
 *   completed(...)        the work finished and the process continues
 *   businessError(...)    a rule was broken, so the process takes its modelled
 *                         BPMN error path (an error catch event). If the model
 *                         has no catch event for the code, Zeebe raises an
 *                         incident, which is still a recorded, visible outcome.
 *   technicalFailure(...) something outside the business rules went wrong;
 *                         the job is retried according to the model's retries
 */

const OUTCOME = {
  COMPLETED: 'completed',
  BUSINESS_ERROR: 'businessError',
  TECHNICAL_FAILURE: 'technicalFailure',
}

/** The job succeeded. `variables` are merged into the process variables. */
function completed(variables = {}) {
  return { kind: OUTCOME.COMPLETED, variables }
}

/**
 * A business rule or the input data stopped the work.
 * `variables` are written before the error is thrown so the process can see
 * what was rejected and route accordingly.
 */
function businessError(errorCode, errorMessage, variables = {}) {
  return { kind: OUTCOME.BUSINESS_ERROR, errorCode, errorMessage, variables }
}

/** An unexpected condition. The job is retried; the process does not move on. */
function technicalFailure(errorMessage, { retries, retryBackOffMs, variables } = {}) {
  return {
    kind: OUTCOME.TECHNICAL_FAILURE,
    errorMessage,
    retries,
    retryBackOffMs,
    variables,
  }
}

/** True when the value is one of the outcomes above. Used to catch handler bugs. */
function isOutcome(value) {
  return Boolean(value) && Object.values(OUTCOME).includes(value.kind)
}

/**
 * Turns an outcome into the job action the SDK expects.
 * The handler must return a job action on every code path, which is why an
 * unrecognised value is reported as a technical failure rather than ignored.
 */
function applyOutcome(job, outcome) {
  switch (outcome.kind) {
    case OUTCOME.COMPLETED:
      return job.complete(outcome.variables)
    case OUTCOME.BUSINESS_ERROR:
      return job.error({
        errorCode: outcome.errorCode,
        errorMessage: outcome.errorMessage,
        variables: outcome.variables,
      })
    case OUTCOME.TECHNICAL_FAILURE: {
      const failure = { errorMessage: outcome.errorMessage }
      if (outcome.retries !== undefined) failure.retries = outcome.retries
      if (outcome.retryBackOffMs !== undefined) failure.retryBackOff = outcome.retryBackOffMs
      if (outcome.variables !== undefined) failure.variables = outcome.variables
      return job.fail(failure)
    }
    default:
      return job.fail({
        errorMessage: `worker handler returned an unrecognised outcome: ${JSON.stringify(outcome)}`,
      })
  }
}

module.exports = {
  OUTCOME,
  completed,
  businessError,
  technicalFailure,
  isOutcome,
  applyOutcome,
}

File: workers/src/main/java/uk/ac/uwe/hprtas/workers/Outcome.java
package uk.ac.uwe.hprtas.workers;

import java.time.Duration;
import java.util.Map;

/**
 * Job outcomes.
 *
 * A worker handler does not touch the job object directly. It returns one of the outcomes below,
 * which keeps the business logic testable on its own and keeps the choice of job action in one
 * place. The three outcomes map onto the three ways a job can end in Camunda 8:
 *
 * <ul>
 *   <li>{@link Completed} - the work finished and the process continues</li>
 *   <li>{@link BusinessError} - a rule was broken, so the process takes its modelled BPMN error
 *       path (an error catch event). If the model has no catch event for the code, Zeebe raises an
 *       incident, which is still a recorded, visible outcome.</li>
 *   <li>{@link TechnicalFailure} - something outside the business rules went wrong; the job is
 *       failed so the broker retries it rather than leaving it stalled</li>
 * </ul>
 *
 * The interface is sealed, so {@link #applyOutcome} is checked at compile time for every outcome a
 * handler can return.
 */
public sealed interface Outcome
    permits Outcome.Completed, Outcome.BusinessError, Outcome.TechnicalFailure {

  /** What the job log records as {@code outcome}. */
  String kind();

  /** The job succeeded. {@code variables} are merged into the process variables. */
  record Completed(Map<String, Object> variables) implements Outcome {

    public Completed {
      variables = Vars.copy(variables);
    }

    @Override
    public String kind() {
      return "completed";
    }
  }

  /**
   * A business rule or the input data stopped the work.
   *
   * The variables are written before the error is thrown. A model has to distinguish the reason for
   * a failure by which catch event was taken rather than by reading one of these variables: the
   * variables attached to a BPMN error are written to the scope of the activity that raised it and
   * are no longer visible in the process scope once the token has moved on.
   */
  record BusinessError(String errorCode, String errorMessage, Map<String, Object> variables)
      implements Outcome {

    public BusinessError {
      variables = Vars.copy(variables);
    }

    @Override
    public String kind() {
      return "businessError";
    }
  }

  /** An unexpected condition. The job is failed; the process does not move on. */
  record TechnicalFailure(
      String errorMessage, Integer retries, Duration retryBackOff, Map<String, Object> variables)
      implements Outcome {

    public TechnicalFailure {
      variables = variables == null ? null : Vars.copy(variables);
    }

    @Override
    public String kind() {
      return "technicalFailure";
    }
  }

  static Completed completed(Map<String, Object> variables) {
    return new Completed(variables);
  }

  static BusinessError businessError(
      ErrorCode errorCode, String errorMessage, Map<String, Object> variables) {
    return new BusinessError(errorCode.name(), errorMessage, variables);
  }

  static TechnicalFailure technicalFailure(String errorMessage) {
    return new TechnicalFailure(errorMessage, null, null, null);
  }

  /** Turns an outcome into the job action the broker is given. */
  static void applyOutcome(JobActions actions, Outcome outcome) {
    switch (outcome) {
      case Completed completed -> actions.complete(completed.variables());

      case BusinessError error ->
          actions.throwError(error.errorCode(), error.errorMessage(), error.variables());

      case TechnicalFailure failure ->
          actions.fail(
              failure.errorMessage(),
              failure.retries(),
              failure.retryBackOff(),
              failure.variables());
    }
  }
}
